File: src/leetcode/m31/NextPermutation.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m31;

/**
 * LeetCode 31
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/11/10 17:12
 */
public class NextPermutation {
    public void nextPermutation(int[] nums) {
        int i = nums.length - 2;

        // Step 1: Find the first decreasing element
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        if (i >= 0) {
            int j = nums.length - 1;
            // Step 2: Find element just larger than nums[i]
            while (nums[j] <= nums[i]) {
                j--;
            }
            // Step 3: Swap nums[i] with nums[j]
            swap(nums, i, j);
        }

        // Step 4: Reverse the subarray from i + 1 to the end
        reverse(nums, i + 1);
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    private void reverse(int[] nums, int start) {
        int end = nums.length - 1;
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }
}

File: src/leetcode/m31/NextPermutation.go
package m31

func nextPermutation(nums []int) {
	n := len(nums)
	n -= 2

	for ; n >= 0; n-- {
		if nums[n] < nums[n+1] {
			break
		}
	}

	if n >= 0 {
		for i := len(nums) - 1; i > n; i-- {
			if nums[i] > nums[n] {
				swap(nums, i, n)
				break
			}
		}
	}

	reverse(nums, n+1)
}

func reverse(nums []int, i int) {
	for j := len(nums) - 1; i < j; i, j = i+1, j-1 {
		swap(nums, i, j)
	}
}

func swap(nums []int, i int, n int) {
	nums[i], nums[n] = nums[n], nums[i]
}
